### PR TITLE
fix: refactor pull flow to use amplify error handling framework

### PR DESCRIPTION
**Pull request by lazpavel**
#### Description of changes
- refactor pull flow to use the new amplify error types and error handling

#### Checklist
- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/commands/publish.ts [hunk 1]
```diff
@@ -12,10 +11,10 @@ export const run = async context => {
...
 
   if (!isHostingAdded) {
-    context.print.info('');
-    context.print.error('Please add hosting to your project before publishing your project');
-    context.print.info('Command: amplify hosting add');
-    context.print.info('');
+    printer.info('');
+    printer.error('Please add hosting to your project before publishing your project');
+    printer.info('Command: amplify hosting add');
+    printer.info('');
```
Comment (jhockett) [inline packages/amplify-cli/src/commands/publish.ts:17]:
```suggestion
    printer.blankLine();
    printer.error('Add hosting to your project before publishing your project');
    printer.info('Command: amplify hosting add');
    printer.blankLine();
```

**Review by jhockett — APPROVED** [review]

**Review by jhockett — DISMISSED** [review]

File: packages/amplify-cli/src/commands/pull.ts [hunk 2]
```diff
@@ -21,8 +20,11 @@ export const run = async (context: $TSContext): Promise<void> => {
     try {
       await preDeployPullBackend(context, inputParams.sandboxId);
     } catch (e) {
-      context.print.error(`Failed to pull sandbox app: ${e.message || 'An unknown error occurred.'}`);
-      showTroubleshootingURL();
+      throw amplifyFaultWithTroubleshootingLink('UnknownFault', {
+        message: `Failed to pull sandbox app: ${e.message || 'An unknown error occurred.'}`,
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-cli/src/commands/pull.ts]: removed the ternary from the `message` and added to the `details` field in case there is no `e.message`

**Review by jhockett — APPROVED** [review]

File: packages/amplify-cli/src/attach-backend.ts [hunk 3]
```diff
@@ -105,22 +112,19 @@ const backupAmplifyFolder = (): void => {
...
-        );
-      }
-      throw e;
+      throw amplifyErrorWithTroubleshootingLink('DirectoryError', {
+        message: `Could not attach the backend to the project.`,
+        resolution: 'Ensure that there are no applications locking the `amplify` folder and try again.',
+        details: e.message,
+        stack: e.stack,
+      });
     }
```
Comment (awsluja) [inline packages/amplify-cli/src/attach-backend.ts:128]: if we only know a resolution for "EPERM", we should keep the IF statement here in the resolution message;
resolution: e.code == 'EPERM' ? "Ensure that..." : "An error occurred attaching the backend.."

**Review by awsluja — CHANGES_REQUESTED** [review]
Looks good overall, my only comment is that we should not eagerly suggest resolution for all errors if we only know one variation of it; this may mislead customers

File: packages/amplify-cli/src/attach-backend.ts (hunk 3 above)
Comment (lazpavel) [reply packages/amplify-cli/src/attach-backend.ts:128]: 'An error occurred attaching the backend' is not a valid resolution, what I can do is to keep the if and throw a fault for anything other that `EPERM`

Comment (awsluja) [reply packages/amplify-cli/src/attach-backend.ts:128]: Ok; that would work. We provide resolution IF we have it; otherwise, throw the fault

**Review by awsluja — APPROVED** [review]
LGTM

**Review by jhockett — APPROVED** [review]

**Pull request merged**